agent/agent.go

```diff
@@ -0,0 +1,431 @@
+	configYAML, err := config.SerializeToYAML(agent.effectiveConfig)
+	if err != nil {
+		agent.logger.Errorf(context.Background(), "Failed to marshal effective config: %v", err)
+		return nil
+	}
+	return &protobufs.EffectiveConfig{
+		ConfigMap: &protobufs.AgentConfigMap{
+			ConfigMap: map[string]*protobufs.AgentConfigFile{
+				"": {Body: []byte(configYAML)},
+			},
+		},
+	}
+}
+
+func (agent *Agent) reportConfigStatus(status protobufs.RemoteConfigStatuses, errorMessage string) {
+	err := agent.opampClient.SetAgentDescription(agent.agentDescription)
+	if err != nil {
+		agent.logger.Errorf(context.Background(), "Could not report OpAMP remote config status: %s", err)
+	}
+	remoteConfigstatus := &protobufs.RemoteConfigStatus{
+		LastRemoteConfigHash: agent.remoteConfig.GetConfigHash(),
+		Status:               status,
+		ErrorMessage:         errorMessage,
+	}
+	err = agent.opampClient.SetRemoteConfigStatus(remoteConfigstatus)
+	if err != nil {
+		agent.logger.Errorf(context.Background(), "Could not report OpAMP remote config status: %s", err)
+		return
+	}
+	agent.remoteConfigStatus = remoteConfigstatus
+}
+
+func (agent *Agent) onMessage(ctx context.Context, msg *types.MessageData) {
+	if msg.OwnMetricsConnSettings != nil {
+		agent.logger.Debugf(ctx, "got own metrics connection settings")
+	}
+	if msg.AgentIdentification != nil {
+		agent.logger.Debugf(ctx, "got agent identification")
+		uid, err := uuid.FromBytes(msg.AgentIdentification.NewInstanceUid)
+		if err != nil {
+			agent.logger.Errorf(ctx, "Failed to parse new instance uid: %v", err)
+			return
+		}
+		agent.updateAgentIdentity(ctx, uid)
+	}
+
+	agent.updateRemoteConfig(ctx, msg)
+
+}
+
+func (agent *Agent) onOpampConnectionSettings(ctx context.Context, settings *protobufs.OpAMPConnectionSettings) error {
+	agent.logger.Debugf(ctx, "got connection settings")
+	return nil
+}
+
+func (agent *Agent) updateRemoteConfig(ctx context.Context, msg *types.MessageData) {
+	if msg.RemoteConfig == nil {
+		agent.logger.Debugf(context.Background(), "updateRemoteConfig: no remote config in message")
+		return
+	}
+
+	if msg.RemoteConfig.GetConfig().GetConfigMap() != nil {
+		// deserialize the config and call ReloadConfig
+		agent.logger.Debugf(ctx, "onMessage got remote config: %v", msg)
+
+		confMap := msg.RemoteConfig.GetConfig().GetConfigMap()
+
+		if !agent.isConfigChanged(msg.RemoteConfig.GetConfigHash()) {
+			agent.logger.Debugf(ctx, "onMessage remote config is the same as the last one, skipping")
+			return
+		}
+
+		var opts []config.ReloadedConfigDataOption
+		if c, ok := confMap["refinery_rules"]; ok {
+			opts = append(opts, config.WithRulesData(config.NewConfigData(c.GetBody(), config.FormatYAML, "opamp://rules")))
+		}
+		if c, ok := confMap["refinery_config"]; ok {
+			opts = append(opts, config.WithConfigData(config.NewConfigData(c.GetBody(), config.FormatYAML, "opamp://config")))
+		}
+
+		agent.remoteConfig = msg.RemoteConfig
```

Comment:
Should we only update agent.remoteConfig if the config reload is successful?

Reply:
Updated `remoteConfig` to `lastRemoteConfigReceived` to clarify its purpose